src/player.cpp

```diff
@@ -10884,42 +10884,45 @@ bool player::has_enough_charges( const item &it, bool show_msg ) const
     return true;
 }
 
-bool player::consume_charges(item *used, long charges_used)
+bool player::consume_charges( item& used, long qty )
 {
-    // Non-tools can use charges too - when they're comestibles
-    const auto tool = dynamic_cast<const it_tool*>(used->type);
-    const auto comest = dynamic_cast<const it_comest*>(used->type);
-    if( charges_used <= 0 || (tool == nullptr && comest == nullptr) ) {
+    if( !has_item( &used ) ) {
+        debugmsg( "Tried to consume charges for an item not in players possession" );
         return false;
     }
 
-    if( tool != nullptr && tool->charges_per_use <= 0 ) {
-        // An item that doesn't normally expend charges is destroyed instead.
-        /* We can't be certain the item is still in the same position,
-         * as other items may have been consumed as well, so remove
-         * the item directly instead of by its position. */
-        i_rem( used );
-        return true;
+    if( qty <= 0 ) {
+        debugmsg( "Tried to consume zero or negagtive charges" );
+        return false;
+    }
+
+    if( !used.is_tool() && !used.is_food() ) {
+        debugmsg( "Tried to consume charges for non-tool, non-food item" );
+        return false;
     }
 
-    if( used->has_flag( "USE_UPS" ) && has_charges( "UPS", tool->charges_per_use ) ) {
-        use_charges( "UPS", charges_used );
-        //Replace 1 with charges it needs to use.
-        if( used->active && used->charges <= 1 && !has_charges( "UPS", 1 ) ) {
-            add_msg_if_player( m_info, _( "You need an UPS of some kind for this %s to work continuously." ), used->tname().c_str() );
+    // Consume comestibles destroying them if no charges remain
+    if( used.is_food() ) {
+        used.charges -= qty;
+        if( used.charges <= 0 ) {
+            i_rem( &used );
+            return true;
         }
-    } else {
-        used->charges -= std::min( used->charges, charges_used );
+        return false;
     }
 
-    if( comest != nullptr && used->charges <= 0 ) {
-        i_rem( used );
+    // Tools which don't require ammo are instead destroyed
+    if( used.is_tool() && !used.ammo_required() ) {
```

Comment:
Wait, I misread the code - it refers to item with 0 charges rather than 0 charge usage.
I think this case may be for items that deploy as something. Manhack for example - it has 0 charges and is a tool.